{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "%%latex\n",
    "\\tableofcontents"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Introduction\n",
    "\n",
    "This notebook will describe the investigation of potential selection trends identified through Hardy-Wienburg Equilibrium (HWE) analysis. Hardy-Wienburg equilibrium describes a set of conditions, under which genotype frequency can be predicted as a product of simpler allele frequencies:\n",
    "\n",
    "$$\n",
    "    p^2+2pq+q^2\n",
    "$$\n",
    "\n",
    "In practice, Hardy-Weinburg is applied through the use of tests such as chi-squared or other exact methods to compare if observed genotype frequencies align with that expected under Hardy-Wienburg Law.\n",
    "\n",
    "We have generated a series of population-subset Hardy-Weinburg reports. These were produced using [Plink-2](https://www.cog-genomics.org/plink/2.0/)'s [`--hardy`](https://www.cog-genomics.org/plink/2.0/basic_stats#hardy) command with the `midp` modifier, in combination with the [`--loop-cats`]() command to loop across population groupings while doing so. These outputs will be used for graphing purposes here.\n",
    "\n",
    "> Refer to the repository and linked documentation site for more information about the [Pharmacogenetics Analysis Pipeline](https://github.com/Tuks-ICMM/Pharmacogenetic-Analysis-Pipeline).\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Objectives"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Notebook configuration"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Dependancies\n",
    "\n",
    "This notebook will make use of the following packages and functions:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [],
   "source": [
    "import plotly.graph_objects as go\n",
    "from plotly.subplots import make_subplots\n",
    "from plotly.io import renderers\n",
    "from pandas import read_csv, Series, cut, qcut, DataFrame, concat\n",
    "from os.path import  join\n",
    "from pathlib import Path\n",
    "from json import load\n",
    "from numpy import sqrt\n",
    "from matplotlib.pyplot import figure,subplots, plot\n",
    "from matplotlib.gridspec import GridSpec\n",
    "import mpltern\n",
    "from seaborn import scatterplot\n",
    "from matplotlib.ticker import MultipleLocator\n",
    "from scipy.interpolate import splrep, BSpline\n",
    "from scipy.interpolate import UnivariateSpline\n",
    "from numpy import linspace"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {
    "tags": [
     "no-input"
    ]
   },
   "outputs": [],
   "source": [
    "renderers.default = \"notebook_connected+pdf\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(join(\"config\", \"manifest copy.json\"), 'r') as config:\n",
    "    config = load(config)\n",
    "\n",
    "path = join(*config[\"output\"])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Data Imports"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We will be using data generated by the [Pharmacogenetics Analysis Pieline](). AS part of thsi analysis, we will need to import and work with the collection of metadata files used to configure this workflow:\n",
    "\n",
    "- `samples.csv`\n",
    "- `locations.csv`\n",
    "- `datasets.csv`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [],
   "source": [
    "SAMPLES = read_csv(join(\"input\", \"samples.csv\"))\n",
    "LOCATIONS = read_csv(join(\"input\", \"locations.csv\"))\n",
    "DATASETS = read_csv(join(\"input\", \"datasets.csv\"))\n",
    "POPULATIONS_TO_ANALYSE = SAMPLES[\"super-population\"].unique().tolist()\n",
    "GENES_TO_ANALYSE = LOCATIONS[\"location_name\"].unique().tolist()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We will also be making use of a VCF-based `MultiIndex` system to organise our hirachical data (`CHROM`, `POS`, `REF` and `ALT`)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [],
   "source": [
    "# First, lets create a dictionary to house all of our graphs.\n",
    "FIGURE = dict()\n",
    "MULTIINDEX = [\"CHROM\", \"POS\", \"REF\", \"ALT\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Type Mapping\n",
    "\n",
    "Our input data can be imported from `.csv` format using the `read_csv()` function from `pandas`. Since this project focuses on more than one gene, we will work our way through the list of genes and import each consolidated report."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [],
   "source": [
    "DATA = dict()\n",
    "for gene in GENES_TO_ANALYSE:\n",
    "    DATA[gene] = read_csv(\n",
    "        join(\n",
    "            path,\n",
    "            \"cleaned\",\n",
    "            f\"super-population_{gene}.csv.zst\",\n",
    "        )\n",
    "        # join(\n",
    "        #     path, \"consolidated_reports\", f\"super-population_{gene}.csv\")\n",
    "            # , sep=\"\\t\"\n",
    "            )\n",
    "    DATA[gene].rename(columns={\"CADD_PHRED\": \"CADD Phred\"}, inplace=True)\n",
    "    DATA[gene].set_index(MULTIINDEX, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "LD_DATA = dict()\n",
    "for gene in GENES_TO_ANALYSE:\n",
    "    LD_DATA[gene] = dict()\n",
    "    for pop in POPULATIONS_TO_ANALYSE:\n",
    "        LD_DATA[gene][pop] = read_csv(join(path, \"linkage_disequilibrium\", \"super-population\", gene, f\"calculated_linkage_disequilibrium_per_cluster.{pop}.vcor.zst\"), sep=\"\\t\")\n",
    "        LD_DATA[gene][pop].rename(columns={\"#CHROM_A\": \"CHROM_A\"}, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "EIGENVECTORS = dict()\n",
    "for gene in GENES_TO_ANALYSE:\n",
    "    _tmp = read_csv(join(path, \"generate_pca\", gene, \"pca.eigenvec\"), sep=\"\\t\")\n",
    "    _tmp.rename(columns={\"#IID\":\"IID\"}, inplace=True)\n",
    "\n",
    "    EIGENVECTORS[gene] = SAMPLES.merge(_tmp, left_on=\"sample_name\", right_on=\"IID\")\n",
    "    EIGENVECTORS[gene].drop([\"IID\"], inplace=True, axis=\"columns\")\n",
    "\n",
    "    del _tmp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "EIGENVALUES = dict()\n",
    "for gene in GENES_TO_ANALYSE:\n",
    "    _tmp = read_csv(join(path, \"generate_pca\", gene, \"pca.eigenvec.allele\"), sep=\"\\t\")\n",
    "    _tmp.rename(columns={\"#CHROM\": \"CHROM\"}, inplace=True)\n",
    "    _tmp.set_index(MULTIINDEX, inplace=True)\n",
    "\n",
    "    EIGENVALUES[gene] = DATA[gene][[\"ID\"]].merge(_tmp, right_index=True, left_index=True)\n",
    "\n",
    "    del _tmp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import missingness consolidation for filtering purposes\n",
    "# CONSOLIDATED_REPORT = read_csv(\n",
    "#     join(\n",
    "#         path,\n",
    "#         \"cleaned\",\n",
    "#         f\"super-population_CYP2B6.csv.zst\",\n",
    "#     )\n",
    "# )\n",
    "\n",
    "# # Set the VEP-based hierarchical `MultiIndex`\n",
    "# CONSOLIDATED_REPORT.set_index(MULTIINDEX, inplace=True)\n",
    "for gene in GENES_TO_ANALYSE:\n",
    "    # For each population in our list of populations to be analyzed...\n",
    "    for population in POPULATIONS_TO_ANALYSE:\n",
    "        # Calculate Heterozygous Frequency from plink2 count data\n",
    "        # DATA[gene][f\"{population} Hetrozygous Freq.\"] = DATA[gene].apply(lambda row: (row[f\"{population}_HET_A1_CT\"]/(row[f\"{population}_HOM_A1_CT\"]+row[f\"{population}_HET_A1_CT\"]+row[f\"{population}_TWO_AX_CT\"]))*100, axis=\"columns\")\n",
    "        # DATA[gene][f\"{population} Hetrozygous Freq.\"] = DATA[gene].apply(lambda row: convertToPercent(row, f\"{population}_HET_A1_CT\", population), axis=\"columns\")\n",
    "        \n",
    "        DATA[gene][f\"{population} Hetrozygous Freq.\"] = (\n",
    "            DATA[gene][f\"{population}_HET_A1_CT\"]\n",
    "            / (\n",
    "                DATA[gene][f\"{population}_HOM_A1_CT\"]\n",
    "                + DATA[gene][f\"{population}_HET_A1_CT\"]\n",
    "                + DATA[gene][f\"{population}_TWO_AX_CT\"]\n",
    "            )\n",
    "        ) * 100\n",
    "\n",
    "\n",
    "        # Calculate Homozygous Reference Frequency from plink2 count DATA[gene]\n",
    "        # DATA[gene][f\"{population} Homozygous Ref Freq.\"] = DATA[gene].apply(lambda row: (row[f\"{population}_HOM_A1_CT\"]/(row[f\"{population}_HOM_A1_CT\"]+row[f\"{population}_HET_A1_CT\"]+row[f\"{population}_TWO_AX_CT\"]))*100, axis=\"columns\")\n",
    "        # DATA[gene][f\"{population} Homozygous Ref Freq.\"] = DATA[gene].apply(lambda row: convertToPercent(row, f\"{population}_HOM_A1_CT\", population), axis=\"columns\")\n",
    "        DATA[gene][f\"{population} Homozygous Ref Freq.\"] = (\n",
    "            DATA[gene][f\"{population}_HOM_A1_CT\"]\n",
    "            / (\n",
    "                DATA[gene][f\"{population}_HOM_A1_CT\"]\n",
    "                + DATA[gene][f\"{population}_HET_A1_CT\"]\n",
    "                + DATA[gene][f\"{population}_TWO_AX_CT\"]\n",
    "            )\n",
    "        ) * 100\n",
    "\n",
    "        # Calculate Homozygous Alternate Frequency from plink2 count data\n",
    "        # DATA[gene][f\"{population} Homozygous Alt Freq.\"] = DATA[gene].apply(lambda row: (row[f\"{population}_TWO_AX_CT\"]/(row[f\"{population}_HOM_A1_CT\"]+row[f\"{population}_HET_A1_CT\"]+row[f\"{population}_TWO_AX_CT\"]))*100, axis=\"columns\")\n",
    "        \n",
    "        # DATA[gene][f\"{population} Homozygous Alt Freq.\"] = DATA[gene].apply(lambda row: convertToPercent(row, f\"{population}_TWO_AX_CT\", population), axis=\"columns\")\n",
    "        \n",
    "        DATA[gene][f\"{population} Homozygous Alt Freq.\"] = (\n",
    "            DATA[gene][f\"{population}_TWO_AX_CT\"]\n",
    "            / (\n",
    "                DATA[gene][f\"{population}_HOM_A1_CT\"]\n",
    "                + DATA[gene][f\"{population}_HET_A1_CT\"]\n",
    "                + DATA[gene][f\"{population}_TWO_AX_CT\"]\n",
    "            )\n",
    "        ) * 100"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for gene in GENES_TO_ANALYSE:\n",
    "    # For each population in our list of populations to be analyzed...\n",
    "    for population in POPULATIONS_TO_ANALYSE:\n",
    "        # Calculate Heterozygous Frequency from plink2 count data\n",
    "        DATA[gene][f\"{population} Hetrozygous Freq.\"] = (\n",
    "            DATA[gene][f\"{population}_HET_A1_CT\"]\n",
    "            / (\n",
    "                DATA[gene][f\"{population}_HOM_A1_CT\"]\n",
    "                + DATA[gene][f\"{population}_HET_A1_CT\"]\n",
    "                + DATA[gene][f\"{population}_TWO_AX_CT\"]\n",
    "            )\n",
    "        ) * 100\n",
    "\n",
    "\n",
    "        # Calculate Homozygous Reference Frequency from plink2 count DATA[gene]\n",
    "        DATA[gene][f\"{population} Homozygous Ref Freq.\"] = (\n",
    "            DATA[gene][f\"{population}_HOM_A1_CT\"]\n",
    "            / (\n",
    "                DATA[gene][f\"{population}_HOM_A1_CT\"]\n",
    "                + DATA[gene][f\"{population}_HET_A1_CT\"]\n",
    "                + DATA[gene][f\"{population}_TWO_AX_CT\"]\n",
    "            )\n",
    "        ) * 100\n",
    "\n",
    "        # Calculate Homozygous Alternate Frequency from plink2 count data\n",
    "        DATA[gene][f\"{population} Homozygous Alt Freq.\"] = (\n",
    "            DATA[gene][f\"{population}_TWO_AX_CT\"]\n",
    "            / (\n",
    "                DATA[gene][f\"{population}_HOM_A1_CT\"]\n",
    "                + DATA[gene][f\"{population}_HET_A1_CT\"]\n",
    "                + DATA[gene][f\"{population}_TWO_AX_CT\"]\n",
    "            )\n",
    "        ) * 100"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Population Structure Figure\n",
    "\n",
    "To start the construction of our figure, we first need to define a `GridSpec`. This object will be used to plot data into a blank figure using a grid-system. Using this method, we can effectively build complex figures with sub-plots of differing sizes and complex layouts."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Initialize Figure"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "GRID_SPEC = dict()\n",
    "\n",
    "for gene in GENES_TO_ANALYSE:\n",
    "    # [CREATE] blank figure\n",
    "    FIGURE[f\"{gene} - Population Structure\"] = figure(figsize=(14, 9))\n",
    "\n",
    "    # [SET] layout algorithm used to optimize space-use.\n",
    "    FIGURE[f\"{gene} - Population Structure\"].set_layout_engine(\"constrained\")\n",
    "\n",
    "    # [SET] Figure title\n",
    "    FIGURE[f\"{gene} - Population Structure\"].suptitle(f\"Population Structure | {gene}\", size=\"x-large\")\n",
    "\n",
    "    # [CREATE] GridSpec with 8 rows and 12 columns\n",
    "    GRID_SPEC[gene] = FIGURE[f\"{gene} - Population Structure\"].add_gridspec(nrows=12, ncols=12)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Principle Component Analysis Sub-Plot"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Next, we need to plot our LD results onto this figure. For this, I want the figure to be on the left-hand side, and occupy a height of $\\frac{5}{8}$ of the available 8 rows (i.e. 5 rows high)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for gene in GENES_TO_ANALYSE:  \n",
    "    # [CREATE] a sub-figure\n",
    "    PCA_SUPER = FIGURE[f\"{gene} - Population Structure\"].add_subfigure(GRID_SPEC[gene][8:12, 0:4])\n",
    "\n",
    "    # [SET] the title of the sub-figure\n",
    "    PCA_SUPER.suptitle(\"PCA | Super-Population\", y=0.85)\n",
    "\n",
    "    # [CREATE] a blank set of axes to plot onto\n",
    "    PCA_SUPER_AX = PCA_SUPER.add_subplot()\n",
    "    PCA_SUPER_AX.tick_params(axis='y', rotation=45, labelsize='x-small')\n",
    "    PCA_SUPER_AX.tick_params(axis='x', labelsize='x-small')\n",
    "    # PCA_SUPER_AX.legend(bbox_to_anchor=(0, 1.2), ncols=5)\n",
    "\n",
    "    # PCA_SUPER_AX.set_position([PCA_SUPER_AX.get_position().x0, PCA_SUPER_AX.get_position().y0, PCA_SUPER_AX.get_position().width, PCA_SUPER_AX.get_position().height])\n",
    "    scatterplot(data=EIGENVECTORS[gene], x=\"PC1\", y=\"PC2\", hue=\"super-population\", ax=PCA_SUPER_AX)\n",
    "    PCA_SUPER_AX.legend(loc='upper center', bbox_to_anchor=(0.5, 1.5), fancybox=True, shadow=True, ncol=5)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for gene in GENES_TO_ANALYSE:  \n",
    "    # [CREATE] a sub-figure\n",
    "    PCA_SEX = FIGURE[f\"{gene} - Population Structure\"].add_subfigure(GRID_SPEC[gene][8:12, 4:8])\n",
    "\n",
    "    # [SET] the title of the sub-figure\n",
    "    PCA_SEX.suptitle(\"PCA | Sex\", y=0.85)\n",
    "\n",
    "    # [CREATE] a blank set of axes to plot onto\n",
    "    PCA_SEX_AX = PCA_SEX.add_subplot()\n",
    "    PCA_SEX_AX.tick_params(axis='y', rotation=45, labelsize='x-small')\n",
    "    PCA_SEX_AX.tick_params(axis='x', labelsize='x-small')\n",
    "    # PCA_SEX_AX.legend(bbox_to_anchor=(0, 1.2), ncols=5)\n",
    "    # PCA_SEX_AX.set_position([PCA_SEX_AX.get_position().x0, PCA_SEX_AX.get_position().y0, PCA_SEX_AX.get_position().width, PCA_SEX_AX.get_position().height])\n",
    "\n",
    "    scatterplot(data=EIGENVECTORS[gene], x=\"PC1\", y=\"PC2\", hue=\"sex\", ax=PCA_SEX_AX)\n",
    "    PCA_SEX_AX.legend(loc='upper center', bbox_to_anchor=(0.5, 1.5), fancybox=True, shadow=True, ncol=5)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for gene in GENES_TO_ANALYSE:  \n",
    "    # [CREATE] a sub-figure\n",
    "    PCA_DATASET = FIGURE[f\"{gene} - Population Structure\"].add_subfigure(GRID_SPEC[gene][8:12, 8:12])\n",
    "\n",
    "    # [SET] the title of the sub-figure\n",
    "    PCA_DATASET.suptitle(\"PCA | Dataset\", y=0.85)\n",
    "\n",
    "    # [CREATE] a blank set of axes to plot onto\n",
    "    PCA_DATASET_AX = PCA_DATASET.add_subplot()\n",
    "    PCA_DATASET_AX.tick_params(axis='y', rotation=45, labelsize='x-small')\n",
    "    PCA_DATASET_AX.tick_params(axis='x', labelsize='x-small')\n",
    "    # PCA_DATASET_AX.legend(bbox_to_anchor=(0, 1.2), ncols=5)\n",
    "    # PCA_DATASET_AX.set_position([PCA_DATASET_AX.get_position().x0, PCA_DATASET_AX.get_position().y0, PCA_DATASET_AX.get_position().width, PCA_DATASET_AX.get_position().height])\n",
    "    scatterplot(data=EIGENVECTORS[gene], x=\"PC1\", y=\"PC2\", hue=\"dataset\", ax=PCA_DATASET_AX)\n",
    "    PCA_DATASET_AX.legend(loc='upper center', bbox_to_anchor=(0.5, 1.5), fancybox=True, shadow=True, ncol=5)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Linkage-Decay Sub-Plot"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Next, we need to plot our LD results onto this figure. For this, I want the figure to be on the left-hand side, and occupy a height of $\\frac{3}{8}$ of the available 8 rows (i.e. 3 rows high)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def binData(data: Series, bin_size: int = 500) -> list:\n",
    "    \"\"\"Function takes a pandas `Series` and generates a categorical bin annotation suitable for aggregation, E.g. `.groupby()`.\n",
    "\n",
    "    Args:\n",
    "        data (Series): A `Series` containing  numerical data to be binned.\n",
    "        bin_size (int, optional): The size of the bins to be made. Defaults to 500.\n",
    "\n",
    "    Returns:\n",
    "        Series: The resulting values in a `Series` object.\n",
    "    \"\"\"\n",
    "    _nearest_multiple_of_bin_size = (round(data.max() / bin_size)) * bin_size\n",
    "\n",
    "    if (data.max() % bin_size) > 0:\n",
    "        _nearest_multiple_of_bin_size += bin_size\n",
    "\n",
    "    _list_of_bins = range(0, _nearest_multiple_of_bin_size + bin_size, bin_size)\n",
    "    _list_of_labels = range(0, _nearest_multiple_of_bin_size, bin_size) # Must be one less than the bins\n",
    "\n",
    "    return cut(\n",
    "           data, bins=_list_of_bins, labels=_list_of_labels\n",
    "        ).to_list()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "LD_DECAY = dict()\n",
    "\n",
    "for gene in GENES_TO_ANALYSE:\n",
    "    # Create a list to store each set of population-level LD results\n",
    "    CONCAT_LIST = list()\n",
    "\n",
    "\n",
    "    # For each population in our list of populations for comparison that were successfully imported\n",
    "    for pop in POPULATIONS_TO_ANALYSE:\n",
    "        # [CALCULATE] distance between pairwise alleles in LD comparison\n",
    "        LD_DATA[gene][pop][\"Distance\"] = abs(\n",
    "                LD_DATA[gene][pop][\"POS_B\"] - LD_DATA[gene][pop][\"POS_A\"]\n",
    "        )\n",
    "\n",
    "        # [CALCULATE] binned distance using a default of 500 base-pairs\n",
    "        LD_DATA[gene][pop][\"Distance (Binned)\"] = binData(LD_DATA[gene][pop][\"Distance\"])\n",
    "\n",
    "        # Here we calculate percentile rankings based on the R^2 values which will be used \n",
    "        # later on to calculate splines for each quantile per population.\n",
    "        LD_DATA[gene][pop][\"Quantile\"] = qcut(\n",
    "            LD_DATA[gene][pop][\"UNPHASED_R2\"],\n",
    "            [0,.25,.5,.75,1],\n",
    "            labels=False,\n",
    "            duplicates=\"drop\"\n",
    "        )\n",
    "\n",
    "        # Drop these unneeded columns\n",
    "        # LD_DATA[gene][pop].drop(\n",
    "        #     columns=[\n",
    "        #         \"CHROM_A\",\n",
    "        #         \"POS_A\",\n",
    "        #         \"CHROM_B\",\n",
    "        #         \"POS_B\",\n",
    "        #         \"ID_A\",\n",
    "        #         \"ID_B\",\n",
    "        #         \"REF_A\",\n",
    "        #         \"REF_B\",\n",
    "        #         \"ALT_A\",\n",
    "        #         \"ALT_B\",\n",
    "        #         \"PROVISIONAL_REF_A?\",\n",
    "        #         \"MAJ_A\",\n",
    "        #         \"PROVISIONAL_REF_B?\",\n",
    "        #         \"MAJ_B\",\n",
    "        #     ],\n",
    "        #     inplace=True,\n",
    "        # )\n",
    "\n",
    "        \n",
    "        \n",
    "        # Here, we assign a categorical label for this batch of linkage results.\n",
    "        # This will be needed later for graphing when we combine all our reports together.\n",
    "\n",
    "        LONG_FORMAT = LD_DATA[gene][pop][[\"UNPHASED_R2\", \"Distance\", \"Distance (Binned)\", \"Quantile\"]].groupby([\"Distance (Binned)\", \"Quantile\"], observed=True).mean().reset_index()\n",
    "        \n",
    "        LONG_FORMAT[\"Population\"] = pop\n",
    "        # Add this populations LD report to the list\n",
    "        CONCAT_LIST.append(LONG_FORMAT)\n",
    "\n",
    "    # Concatenate the list of population-level LD reports for this gene together\n",
    "    LD_DECAY[gene] = concat(CONCAT_LIST)\n",
    "\n",
    "    # # [CREATE] a sub-figure\n",
    "    # LD = FIGURE[f\"{gene} - Population Structure\"].add_subfigure(GRID_SPEC[gene][0:8, 0:4])\n",
    "\n",
    "    # # [SET] the title of the sub-figure\n",
    "    # LD.suptitle(\"LD Decay | Chr Band\")\n",
    "\n",
    "    # TMP_GRIDSPEC = LD.add_gridspec(nrows=len(PLOT_LD_DATA[\"Quantile\"].unique()), ncols=1)\n",
    "\n",
    "    # for quantileIndex, quantile in enumerate(PLOT_LD_DATA[\"Quantile\"].unique()):\n",
    "            \n",
    "    #     # [CREATE] a blank set of axes to plot onto\n",
    "    #     LD_AX = LD.add_subplot(TMP_GRIDSPEC[quantileIndex,:])\n",
    "\n",
    "    #     pos = LD_AX.get_position()\n",
    "    #     LD_AX.set_position([pos.x0, pos.y0 + (pos.height * 0.05), pos.width, pos.height * 0.95])\n",
    "\n",
    "    #     LD_AX.set_ylim(0.2,1)\n",
    "    #     LD_AX.set_xlim(0,PLOT_LD_DATA[\"Distance (Binned)\"].max())\n",
    "\n",
    "    #     LD_AX.tick_params(axis='y', rotation=45, labelsize='x-small')\n",
    "    #     LD_AX.tick_params(axis='x', labelsize='x-small')\n",
    "\n",
    "    #     TMP = PLOT_LD_DATA.loc[PLOT_LD_DATA[\"Quantile\"] == quantile]\n",
    "\n",
    "    #     # [CREATE] line-plot of LD Decay results | TODO: EXPAND\n",
    "    #     scatterplot(TMP, x=\"Distance (Binned)\", y=\"UNPHASED_R2\", hue=\"Population\", ax=LD_AX, s=1, legend=False)\n",
    "\n",
    "\n",
    "    #     del PLOT_LD_DATA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for gene in GENES_TO_ANALYSE:\n",
    "    # [CREATE] a sub-figure\n",
    "    LD = FIGURE[f\"{gene} - Population Structure\"].add_subfigure(GRID_SPEC[gene][0:8, 0:4])\n",
    "\n",
    "    # [SET] the title of the sub-figure\n",
    "    LD.suptitle(\"LD Decay | Chr Band\")\n",
    "\n",
    "    LD.supylabel(\"$R^2$ (Unphased)\", size=\"small\")\n",
    "\n",
    "    TMP_GRIDSPEC = LD.add_gridspec(nrows=len(LD_DECAY[gene][\"Quantile\"].unique()), ncols=1)\n",
    "\n",
    "    for quantileIndex, quantile in enumerate(LD_DECAY[gene][\"Quantile\"].unique()):\n",
    "            \n",
    "        # [CREATE] a blank set of axes to plot onto\n",
    "        LD_AX = LD.add_subplot(TMP_GRIDSPEC[quantileIndex,:])\n",
    "\n",
    "        pos = LD_AX.get_position()\n",
    "        LD_AX.set_position([pos.x0, pos.y0 + (pos.height * 0.05), pos.width, pos.height * 0.95])\n",
    "\n",
    "        # LD_AX.set_ylim(0.2,1)\n",
    "        LD_AX.set_xlim(0,LD_DECAY[gene][\"Distance (Binned)\"].max())\n",
    "\n",
    "        LD_AX.tick_params(axis='y', rotation=45, labelsize='x-small')\n",
    "        LD_AX.tick_params(axis='x', labelsize='x-small')\n",
    "\n",
    "        TMP = LD_DECAY[gene].loc[LD_DECAY[gene][\"Quantile\"] == quantile]\n",
    "\n",
    "        # [CREATE] line-plot of LD Decay results | TODO: EXPAND\n",
    "        scatterplot(TMP, x=\"Distance (Binned)\", y=\"UNPHASED_R2\", hue=\"Population\", ax=LD_AX, s=1, legend=False)\n",
    "\n",
    "        LD_AX.set_ylabel(\" \")\n",
    "\n",
    "        AFR = TMP.loc[TMP[\"Population\"] == \"AFR\"]\n",
    "\n",
    "        spline = UnivariateSpline(AFR[\"Distance (Binned)\"], AFR[\"UNPHASED_R2\"], s=0.75, k=3)\n",
    "\n",
    "        x_smooth = linspace(AFR[\"Distance (Binned)\"].min(), AFR[\"Distance (Binned)\"].max(), 500)\n",
    "        y_smooth = spline(x_smooth)\n",
    "\n",
    "        LD_AX.plot(x_smooth, y_smooth, linestyle=\"dashdot\", linewidth=2)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for gene in GENES_TO_ANALYSE:\n",
    "    \n",
    "    # [CREATE] a sub-figure\n",
    "    HWE = FIGURE[f\"{gene} - Population Structure\"].add_subfigure(GRID_SPEC[gene][0:8, 4:12])\n",
    "\n",
    "    # [CREATE] a gridspec to layout multiple sub-plots in this sub-figure\n",
    "    HWE_GRIDSPEC = HWE.add_gridspec(nrows=2, ncols=4)\n",
    "\n",
    "    # [SET] the title of the sub-figure\n",
    "    HWE.suptitle(f\"HWE | Gene Region\")\n",
    "\n",
    "    for pop in POPULATIONS_TO_ANALYSE:\n",
    "        # Calculate the grid position this plot should be placed under\n",
    "        # First, calculate the population position in the flat list\n",
    "        _position = POPULATIONS_TO_ANALYSE.index(pop)\n",
    "        # Next, divide by the number of columns and store the division result (i.e. row number) and remainder\n",
    "        _row, _col = divmod(_position, 4)\n",
    "            \n",
    "        # Identify SNPs with non-significant deviations from HWE\n",
    "        NOT_SIGNIFICANTLY_DIFFERENT = DATA[gene].loc[\n",
    "            DATA[gene][f\"{pop}_MIDP\"] > 0.5\n",
    "        ]\n",
    "\n",
    "        # Identify SNPs with significant deviations from HWE\n",
    "        SIGNIFICANTLY_DIFFERENT = DATA[gene].loc[\n",
    "            DATA[gene][f\"{pop}_MIDP\"] <= 0.5\n",
    "        ]\n",
    "\n",
    "        TERNARY_FIGURE = HWE.add_subfigure(HWE_GRIDSPEC[_row, _col])\n",
    "        TERNARY_FIGURE.suptitle(pop, size=\"x-small\")\n",
    "        TERNARY_FIGURE_AXES = TERNARY_FIGURE.add_subplot(projection=\"ternary\")\n",
    "        TERNARY_FIGURE_AXES.scatter(\n",
    "            NOT_SIGNIFICANTLY_DIFFERENT[f\"{pop} Hetrozygous Freq.\"], \n",
    "            NOT_SIGNIFICANTLY_DIFFERENT[f\"{pop} Homozygous Ref Freq.\"], \n",
    "            NOT_SIGNIFICANTLY_DIFFERENT[f\"{pop} Homozygous Alt Freq.\"], \n",
    "            color=\"darkgray\",\n",
    "            edgecolors=\"white\", \n",
    "            linewidths=0.5,\n",
    "            zorder=2,\n",
    "            s=10\n",
    "            )\n",
    "        TERNARY_FIGURE_AXES.scatter(\n",
    "            SIGNIFICANTLY_DIFFERENT[f\"{pop} Hetrozygous Freq.\"], \n",
    "            SIGNIFICANTLY_DIFFERENT[f\"{pop} Homozygous Ref Freq.\"], \n",
    "            SIGNIFICANTLY_DIFFERENT[f\"{pop} Homozygous Alt Freq.\"],\n",
    "            color=\"red\" ,\n",
    "            edgecolors=\"white\", \n",
    "            linewidths=0.5, \n",
    "            zorder=1,\n",
    "            marker=\"X\",\n",
    "            s=50\n",
    "            )\n",
    "        TERNARY_FIGURE_AXES.tribin(\n",
    "            SIGNIFICANTLY_DIFFERENT[f\"{pop} Hetrozygous Freq.\"], \n",
    "            SIGNIFICANTLY_DIFFERENT[f\"{pop} Homozygous Ref Freq.\"], \n",
    "            SIGNIFICANTLY_DIFFERENT[f\"{pop} Homozygous Alt Freq.\"],\n",
    "            gridsize=5,\n",
    "            edgecolors=\"white\", \n",
    "            linewidths=0.5,\n",
    "            bins=\"log\",\n",
    "            zorder=0\n",
    "            )\n",
    "        \n",
    "        TERNARY_FIGURE_AXES.set_tlabel(\"AA\", size=\"small\")\n",
    "        TERNARY_FIGURE_AXES.set_llabel(\"AB\", size=\"small\")\n",
    "        TERNARY_FIGURE_AXES.set_rlabel(\"BB\", size=\"small\")\n",
    "        TERNARY_FIGURE_AXES.taxis.set_major_locator(MultipleLocator(0.2))\n",
    "        TERNARY_FIGURE_AXES.laxis.set_major_locator(MultipleLocator(0.2))\n",
    "        TERNARY_FIGURE_AXES.raxis.set_major_locator(MultipleLocator(0.2))\n",
    "        TERNARY_FIGURE_AXES.taxis.set_tick_params(\n",
    "            tick2On=True,\n",
    "            colors=\"C0\",\n",
    "            grid_color=\"C0\",\n",
    "            labelsize=6,\n",
    "            gridOn=True,\n",
    "            grid_alpha=0.45,\n",
    "            zorder=0.0\n",
    "            )\n",
    "        TERNARY_FIGURE_AXES.laxis.set_tick_params(\n",
    "            tick2On=True,\n",
    "            colors=\"C1\",\n",
    "            grid_color=\"C1\",\n",
    "            labelsize=6,\n",
    "            gridOn=True,\n",
    "            grid_alpha=0.45,\n",
    "            zorder=0.0\n",
    "            )\n",
    "        TERNARY_FIGURE_AXES.raxis.set_tick_params(\n",
    "            tick2On=True,\n",
    "            colors=\"C2\",\n",
    "            grid_color=\"C2\",\n",
    "            labelsize=6,\n",
    "            gridOn=True,\n",
    "            grid_alpha=0.45,\n",
    "            zorder=0.0\n",
    "            )\n",
    "        TERNARY_FIGURE_AXES.taxis.label.set_color(\"C0\")\n",
    "        TERNARY_FIGURE_AXES.laxis.label.set_color(\"C1\")\n",
    "        TERNARY_FIGURE_AXES.raxis.label.set_color(\"C2\")\n",
    "        TERNARY_FIGURE_AXES.spines[\"tside\"].set_color(\"C0\")\n",
    "        TERNARY_FIGURE_AXES.spines[\"lside\"].set_color(\"C1\")\n",
    "        TERNARY_FIGURE_AXES.spines[\"rside\"].set_color(\"C2\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for gene in GENES_TO_ANALYSE:\n",
    "    display(FIGURE[f\"{gene} - Population Structure\"])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Hardy-Weinberg Equilibrium Ternary Plots"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Plotting Objectievs\n",
    "\n",
    "- HWE\n",
    "    - Significant vs Non-Significant\n",
    "    - Significance density\n",
    "- PCA\n",
    "    - Magnitude of contribution towards population structure\n",
    "    - Direction of contribution towards population structure"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Next, I want to visualize the mutations with significant deviations from a normal HWE distribution. I also however, want to highlight the density of these significant mutations, as well as highlight individual mutations which contribute strongly towards our observed population structure."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To accomplish density highlighting, we can make use of the `tribin` plotting method by `matplotlib`, and the ternary projections (axes) provided by `mpltern`. This will add a color-coded tri-bin plot that highlights bin density. This can be placed underneath our data-points with a reduced opacity to provive clearer insight into scatterplot point-density."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### PCA Scores and HWE point size"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To facilitate the identification of mutations which contribute strongly towards population-structure, as seen in the PCA, we can make use of the scatterplot's point-sizes and vary this property per-mutation based on the mutations PCA scores. This will make mutations which contribute strongly towards the observed PCA structure larger."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We quickly just need to scale the allele-weights retrieved from teh PCA report. This is because the scores are positive or negative, and can be very small or very large. A naive approach may be to perform basic Min-Max scaling, however due to the distribution of points, some data points will come out so small they wont be legible. For this reason, we will be scaling teh column to standard values:\n",
    "\n",
    "$$\n",
    "Scaled Value = \\frac{Original Value - Min Value}{Max Value - Min Value}*100\n",
    "$$\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Scaling PCA Allele-Scores to a Fixed Range\n",
    "\n",
    "We first need to scale the PCA allele-scores obtained from Plink-2's `--pca allele-wts` function to the fixed range $[0, 10]$. This is because we need to provide a number in the range of $[0, \\text{inf}]$ to the `scatter` function responsible for controlling point-size (Unfortunately it does not have a magical internal method for doing this already,a s this is a somewhat specialized graphing application)\n",
    "\n",
    "#### Step-by-Step Explanation\n",
    "\n",
    "1. **Input Range**:\n",
    "   Let the original PCA allele-scores be in the range $[x_{\\text{min}}, x_{\\text{max}}]$, where:\n",
    "   - $|x_{\\text{min}}|$ = The absolute value of the minimum score\n",
    "   - $|x_{\\text{max}}|$ = The absolute value of the minimum score\n",
    "\n",
    "2. **Target Range**:\n",
    "   The desired scaled range is $[y_{\\text{min}}, y_{\\text{max}}]$, where:\n",
    "   - $y_{\\text{min}} = 0$\n",
    "   - $y_{\\text{max}} = 10$\n",
    "\n",
    "3. **Scaling Transformation**:\n",
    "   To map an original value $x$ to the scaled value $y$, we use the linear transformation formula:\n",
    "   $$\n",
    "   y = \\frac{x - x_{\\text{min}}}{x_{\\text{max}} - x_{\\text{min}}} \\times (y_{\\text{max}} - y_{\\text{min}}) + y_{\\text{min}}\n",
    "   $$\n",
    "\n",
    "4. **Simplified for Target Range $[0, 10]$**:\n",
    "   Substituting $y_{\\text{min}} = 0$ and $y_{\\text{max}} = 10$, the formula becomes:\n",
    "   $$\n",
    "   \\text{Scaled Value} = \\frac{x - x_{\\text{min}}}{x_{\\text{max}} - x_{\\text{min}}} \\times 10\n",
    "   $$\n",
    "\n",
    "#### Key Notes\n",
    "- This formula ensures that the smallest value in the original range ($x_{\\text{min}}$) maps to 0, and the largest value ($x_{\\text{max}}$) maps to 10.\n",
    "- The transformation is linear, preserving the relative differences between values in the original range.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Hardy-Weinberg Results"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Calculate Genotype frequency\n",
    "\n",
    "The results contain genotype counts, so our first step is to convert these into genotype frequencies that can be plotted on a ternary plot. To do this efficiently, we will loop through each gene found in the `DataFrame` under our chosen cluster of '_super-population_'. For each `DataFrame`, we will then loop again through each unique population annotated in our samples files under '_super-population_' to calculate sub-group genotype frequencies."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Generate ternary sub-plots\n",
    "\n",
    "Next, we will make a composite plot using the plotly `make_subplots` function which allows the creation of a figure containing multiple different sub-plots. Since there are an even number of gene regions in this analysis, we can simply create a grid of 2x4.\n",
    "\n",
    "### Graph type\n",
    "\n",
    "This plot will be slightly more complicated to construct since we want to style and add multiple subsets of data under different ledgend entries. This will allow us to more easily distinguish between variants which diverge from HWE and those that do not via plot styling. Normally, this would be facilitated by long-format data and the `plotly.express` interface for constructing `plotly` `Figure`s. In this case, there are some operations we will not be able to perform through this interface, which means we will have to construct each 'trace' manually via the `plotly.graph_objects` interface for Object-oriented plotting.\n",
    "\n",
    "### Subplot grid\n",
    "\n",
    "Since this plot will involve multiple populations in a sub-plot format, a system will be needed to position each ternary-plot within each overall gene-level HWE `Figure`. Fortunately, determining the grid-position of a population in a list of known length can be acheived through mathematical operations and operators already available in python. Provided we know the maximum number of graphs on one row of the plot, we can make use of the `divmod()` which performs division (`/` operator) and modulus (`%` operator) calculations to determine row-grid coordinates from population index position. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We quickly just need to scale the allele-weights retrieved from teh PCA report. This is because the scores are positive or negative, and can be very small or very large. A naive approach may be to perform basic Min-Max scaling, however due to the distribution of points, some data points will come out so small they wont be legible. For this reason, we will be scaling teh column to standard values:\n",
    "\n",
    "$$\n",
    "Scaled Value = \\frac{Original Value - Min Value}{Max Value - Min Value}*100\n",
    "$$\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Export"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now we just need to export these figures to file for writing purposes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "Path(join(path, \"Graphs\", \"05\")).mkdir(exist_ok=True, parents=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for figure in FIGURE.keys():\n",
    "    FIGURE[figure].savefig(\n",
    "        join(path, \"Graphs\", \"05\", f\"{figure}.jpeg\"),\n",
    "        dpi=300\n",
    "    )"
   ]
  }
 ],
 "metadata": {
  "authors": [
   {
    "name": "Graeme R. Ford"
   }
  ],
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
